### PR TITLE
Add Fusion 360 tool library import

### LINKED ISSUE #7: Adding Vendors catalogs
I was looking for your import file format specs, and don't see them. (I did look under docs)

I would like to bring in catalogs from Whiteside & Bits&Bits.

Is there a JSON or such format you prefer? ( I already can make FreeCad tools and libraries but your BTL stuff is a mystery)

I know in a previous chat I said I was using CSV or Excel from vendors, because frankly most vendors don't understand anything more than that.

### PR DESCRIPTION
**Pull request by alexneufeld**
closes #7

Example data to test this is available at [https://cam.autodesk.com/hsmtools](https://cam.autodesk.com/hsmtools)

I added custom tool shapes for importing from fusion, because the dimensions and shapes don't translate 1:1 E.G.:
- fusion specifies an additional shoulder length parameter for all tools that isn't present in most of the btl shapes. 
- dovetail cutter have rounded edges instead of truncated
- some tools may have rounded or square edges, so the shape .fcstd file need to swap the revolved sketch out on the fly

Perhaps we can figure out a way to hide these import specific shapes from the btl add tool gui?

image

There is also a performance issue somewhere - a 15Mb json library takes ~5 minutes to import.

This is definitely still in a draft state. Have a look, see what you think.

### CODE REVIEW

**Comment by knipknap** [comment]
Wow, that looks crazy awesome, great work!

Hopefully the performance issue is not caused by the renderer, because that could be hard to fix.

**Comment by alexneufeld** [comment]
> Wow, that looks crazy awesome, great work!
> 
> Hopefully the performance issue is not caused by the renderer, because that could be hard to fix.

Doesn't look like the renderer is the issue - `btl/library.py` just isn't efficient enough to handle ~1000 tools. Here are my preliminary optimization results:

test data: [https://cam.autodesk.com/tools/download.php?name=Harvey%20Tool/End%20Mills](https://cam.autodesk.com/tools/download.php?name=Harvey%20Tool/End%20Mills)

test script:

``` python
# run with freecad -c ~/test_script.py >> optimization_logs.txt
import os
import sys
from cProfile import Profile
from pstats import SortKey, Stats
from btl.serializers import FusionJSONSerializer


with Profile() as profile:
    filename = "/path/to/Harvey Tool-End Mills.json"
    dirname = os.path.dirname(filename)
    serializer = FusionJSONSerializer(dirname)
    # Try to read the file.
    library = serializer.deserialize_library_from_file(filename)
    (
        Stats(profile)
        .strip_dirs()
        .sort_stats(SortKey.CALLS)
        .print_stats()
    )

sys.exit(0)

```

trimmed test results:

```
// before optimization
363418835 function calls (363406722 primitive calls) in 123.958 seconds

   Ordered by: call count

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
356188145   62.841    0.000   62.841    0.000 tool.py:34(__eq__)
  1103131    0.164    0.000    0.164    0.000 {built-in method builtins.isinstance}
   493808    0.065    0.000    0.065    0.000 {method 'get' of 'dict' objects}
   304032    0.052    0.000    0.052    0.000 {method 'join' of 'str' objects}
   302660    0.047    0.000    0.047    0.000 {method 'group' of 're.Match' objects}
   302656    0.437    0.000    0.536    0.000 sre_parse.py:1066(expand_template)
   302656    0.116    0.000    0.653    0.000 re.py:330(filter)
   223018    0.033    0.000    0.033    0.000 {built-in method sys._getframe}
223018/222967    0.099    0.000    0.538    0.000 {built-in method builtins.__import__}
223015/222967    0.426    0.000    1.042    0.000 feature.py:105(_import)
...

// after optimization

7111326 function calls (7111149 primitive calls) in 6.957 seconds

   Ordered by: call count

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  1103131    0.154    0.000    0.154    0.000 {built-in method builtins.isinstance}
   481871    0.054    0.000    0.054    0.000 {method 'get' of 'dict' objects}
   304032    0.050    0.000    0.050    0.000 {method 'join' of 'str' objects}
   302660    0.044    0.000    0.044    0.000 {method 'group' of 're.Match' objects}
   302656    0.428    0.000    0.521    0.000 sre_parse.py:1066(expand_template)
   302656    0.119    0.000    0.640    0.000 re.py:330(filter)
   223018    0.030    0.000    0.030    0.000 {built-in method sys._getframe}
223018/222967    0.093    0.000    0.556    0.000 {built-in method builtins.__import__}
223015/222967    0.414    0.000    1.044    0.000 feature.py:105(_import)
...

```

df5502322bbb6c771510d256b1cfde95b969b1a3 is pretty unsophisticated, I just deleted some expensive safety checks in the tool addition process. Looks like storing the tools as a list is particularly bad because of how often we check for membership in the library and or iterate through it:

``` python
class Library(object):
    API_VERSION = 1

    def __init__(self, label, id=None):
        self.id = id or str(uuid.uuid1())
        self.label = label
        self.tools = []  # <-- bad
        ...
```

Any thoughts?

**Comment by knipknap** [comment]
That looks like quite the performance improvement! The tools are already stored in two data structures to improve the performance of lookups, the next line in the constructor is:

        self.tool_nos = {}  # Maps tool_no number to tool

But neither structure allows for easily finding the next available tool number.

When implementing the `assign_new_tool_number()` method, I was trying to avoid using the highest number, because that used to cause some inconvenience in the FreeCAD internal tool library UI before. Some logic was added to avoid "accidental reshuffling" when passing by an already occupied tool number while a user spins through them in the UI, using the spin buttons of the tool number field.

But I think my reasoning has become rather obsolete, since after introducing the pocket attribute, the situations where "just using the highest number" was a problem should no longer exist. At least I can't think of one anymore.

So in short, I think using a tool_no_inc attribute is good.

Just one small nitpick: We have to be very careful with the naming, as it's easy to confuse tool ID, tool number, and tool pocket. So I'd suggest to call the counter tool_no_inc or something.

**Comment by alexneufeld** [comment]
next optimization: replace `QListWidget` and `TwoLineTableCell` with `QTableView` and a custom data model. *Rendering* the tool table view in the library browser is now several orders of magnitude faster:

image

script:

``` python
import os
from btl import ToolDB, serializers


from btl.ui import init
from btl.ui.library import LibraryUI
from cProfile import Profile
from pstats import SortKey, Stats
import FreeCADGui
from pprint import pprint

prefs = FreeCAD.ParamGet("User parameter:BaseApp/Preferences/Mod/Path")



# pprint(widget.__dir__())
# Ensure that a library dir is defined in the preferences.
lib_dir = prefs.GetString("LastPathToolLibrary", "~/.btl/Library")
lib_base_dir = os.path.dirname(lib_dir)
#FreeCAD.Console.PrintMessage("Library path is {}\n".format(lib_base_dir))

# Note: Creating the serializer automatically initializes the library
# directory.
tool_db = ToolDB()
lib_base_dir = os.path.expanduser(lib_base_dir)
serializer = serializers.FCSerializer(lib_base_dir)

dialog = LibraryUI(tool_db, serializer, parent=FreeCADGui.getMainWindow())

with Profile() as profile:
    # Gui.runCommand('Path_ToolBitLibraryOpen',0)
    dialog.form.comboBoxLibrary.setCurrentIndex(7)
    (
        Stats(profile)
        .strip_dirs()
        .sort_stats(SortKey.CALLS)
        .print_stats()
    )

dialog.show()
```

results:

```
// with ListWidget
         19205012 function calls in 31.729 seconds

   Ordered by: call count

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 18045028    3.835    0.000    3.835    0.000 tool.py:34(__eq__)
   150315    2.698    0.000    2.698    0.000 {method 'setText' of 'PySide2.QtWidgets.QLabel' objects}
   120252    0.025    0.000    0.025    0.000 tablecell.py:48(_highlight)
   108132    0.018    0.000    0.018    0.000 {method 'get' of 'dict' objects}
    84098    0.047    0.000    0.047    0.000 {method 'format' of 'str' objects}
    72088    0.008    0.000    0.008    0.000 {built-in method builtins.isinstance}
    48112    0.008    0.000    0.008    0.000 {method 'lower' of 'str' objects}
    ...

// TableView with only enough stuff implemented to show *something*
    24 function calls in 0.000 seconds

   Ordered by: call count

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        3    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        2    0.000    0.000    0.000    0.000 library.py:38(rowCount)
        2    0.000    0.000    0.000    0.000 library.py:41(columnCount)
        1    0.000    0.000    0.000    0.000 cProfile.py:51(create_stats)
        1    0.000    0.000    0.000    0.000 pstats.py:107(__init__)
        1    0.000    0.000    0.000    0.000 pstats.py:117(init)
        1    0.000    0.000    0.000    0.000 pstats.py:136(load_stats)
        1    0.000    0.000    0.000    0.000 library.py:216(update_button_state)
        1    0.000    0.000    0.000    0.000 library.py:253(library_selected)
        1    0.000    0.000    0.000    0.000 library.py:257(get_selected_library)
        1    0.000    0.000    0.000    0.000 library.py:263(update_tool_list)
        1    0.000    0.000    0.000    0.000 library.py:26(__init__)
        1    0.000    0.000    0.000    0.000 {method 'currentData' of 'PySide2.QtWidgets.QComboBox' objects}
        1    0.000    0.000    0.000    0.000 {method 'setCurrentIndex' of 'PySide2.QtWidgets.QComboBox' objects}
        1    0.000    0.000    0.000    0.000 {method 'setModel' of 'PySide2.QtWidgets.QTableView' objects}
        1    0.000    0.000    0.000    0.000 db.py:19(get_library_by_id)
        1    0.000    0.000    0.000    0.000 library.py:42(get_tools)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}

// with all functionality reimplemented using the new table table view
6038 function calls in 0.003 seconds

   Ordered by: call count

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     6007    0.001    0.000    0.001    0.000 {method 'setRowHidden' of 'PySide2.QtWidgets.QTableView' objects}
        5    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        4    0.000    0.000    0.000    0.000 tablemodel.py:36(rowCount)
        2    0.000    0.000    0.000    0.000 tablemodel.py:39(columnCount)
        2    0.000    0.000    0.000    0.000 library.py:179(update_button_state)
        1    0.000    0.000    0.000    0.000 cProfile.py:51(create_stats)
        1    0.000    0.000    0.000    0.000 pstats.py:107(__init__)
        1    0.000    0.000    0.000    0.000 pstats.py:117(init)
        1    0.000    0.000    0.000    0.000 pstats.py:136(load_stats)
        1    0.002    0.002    0.002    0.002 library.py:239(update_search)
        1    0.000    0.000    0.000    0.000 tablemodel.py:7(__init__)
        1    0.000    0.000    0.003    0.003 library.py:216(library_selected)
        1    0.000    0.000    0.000    0.000 library.py:220(get_selected_library)
        1    0.000    0.000    0.003    0.003 library.py:226(update_tool_list)
        1    0.000    0.000    0.000    0.000 {method 'currentData' of 'PySide2.QtWidgets.QComboBox' objects}
        1    0.000    0.000    0.003    0.003 {method 'setCurrentIndex' of 'PySide2.QtWidgets.QComboBox' objects}
        1    0.000    0.000    0.000    0.000 {method 'text' of 'PySide2.QtWidgets.QLineEdit' objects}
        1    0.000    0.000    0.000    0.000 {method 'setModel' of 'PySide2.QtWidgets.QTableView' objects}
        1    0.000    0.000    0.000    0.000 db.py:19(get_library_by_id)
        1    0.000    0.000    0.000    0.000 library.py:42(get_tools)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
```

Still more to do. Also make sure to back up your tool libraries before testing this, just in case I got something wrong...

**Comment by knipknap** [comment]
Looks promising! Looks like table view still isn't quite ready aesthetically (e.g. the rich rendering missing and search/highlighting, etc), which may impact performance a bit.

I would propose to keep the.UI performance improvement separate from the fusion feature, because the PR is getting quite big otherwise.

**Comment by alexneufeld** [comment]
OK, I've trimmed this PR down to just what's needed to import tools from fusion. Note that form tools and thread mills aren't supported yet as they can have much more complex geometry. I can probably add support for these later, but we'll leave that for a future PR.

I changed the import-specific tool shape files' names to start with an underscore, and set them to not appear in the 'Add Tool' user interface. This is based on the assumption that users shouldn't use these shape files if they aren't importing tools from fusion 360. Let me know if that seems like a bad assumption.

**Comment by knipknap** [comment]
Looks great! I agree that hiding the shapes makes sense. Perhaps we can adopt some of the fusion shapes as built-in, for example the lollipop one has no similar equivalent in FreeCAD. But we could decide that later, I don't want to risk incompatibility with FreeCAD Path by introducing a shape that they later introduce in a different manner.

So the PR looks ready to merge, can I merge it?

**Comment by alexneufeld** [comment]
I'm ready to merge if you are

**Comment by knipknap** [comment]
Thanks for the PR!
